libs/core-ui/src/lib/util/JointDataset.ts

```diff
@@ -691,6 +537,246 @@ export class JointDataset {
 
+  private initializeDatasetVariables(args: IJointDatasetArgs): void {
+    if (args.featureMetaData !== undefined) {
+      this.datasetMetaData = { featureMetaData: args.featureMetaData };
+    }
+
+    if (args.dataset && args.dataset.length > 0) {
+      this.initializeDataDictIfNeeded(args.dataset);
+      this.datasetRowCount = args.dataset.length;
+      this.datasetFeatureCount = args.dataset[0].length;
+      // first set metadata
+      args.dataset[0].forEach((_, colIndex) => {
+        const key = JointDataset.DataLabelRoot + colIndex.toString();
+        if (args.metadata.featureIsCategorical?.[colIndex]) {
+          const sortedUnique = (
+            args.metadata.featureRanges[colIndex] as ICategoricalRange
+          ).uniqueValues
+            .concat()
+            .sort();
+          this.metaDict[key] = {
+            abbridgedLabel: args.metadata.featureNamesAbridged[colIndex],
+            category: ColumnCategories.Dataset,
+            index: colIndex,
+            isCategorical: true,
+            label: args.metadata.featureNames[colIndex],
+            sortedCategoricalValues: sortedUnique,
+            treatAsCategorical: true
+          };
+        } else {
+          this.metaDict[key] = {
+            abbridgedLabel: args.metadata.featureNamesAbridged[colIndex],
+            category: ColumnCategories.Dataset,
+            featureRange: args.metadata.featureRanges[
+              colIndex
+            ] as INumericRange,
+            index: colIndex,
+            isCategorical: false,
+            label: args.metadata.featureNames[colIndex]
+          };
+        }
+      });
+      args.dataset.forEach((row, index) => {
+        row.forEach((val, colIndex) => {
+          const key = JointDataset.DataLabelRoot + colIndex.toString();
+          if (this.dataDict) {
+            // store the index for categorical values rather than the actual value. Makes dataset uniform numeric and enables dithering
+            if (args.metadata.featureIsCategorical?.[colIndex]) {
+              this.dataDict[index][key] =
+                this.metaDict[key].sortedCategoricalValues?.indexOf(val) || 0;
+            } else {
+              this.dataDict[index][key] = val;
+            }
+          }
+        });
+      });
+      this.hasDataset = true;
+    }
+  }
+
+  private initializeLabelVariables(
+    args: IJointDatasetArgs,
+    modelTypeIsObjectDetection: boolean
+  ): void {
+    if (args.predictedY && !modelTypeIsObjectDetection) {
+      this.updateMetaDataDict(
+        args.predictedY,
+        args.metadata,
+        JointDataset.PredictedYLabel,
+        localization.Interpret.ExplanationScatter.predictedY,
+        localization.Interpret.ExplanationScatter.predictedY,
+        args.targetColumn
+      );
+      this.hasPredictedY = true;
+    } else if (args.objectDetectionLabels && modelTypeIsObjectDetection) {
+      this.updateMetaDataDict(
+        args.objectDetectionLabels.map((label) => label.incorrect),
```

Comment:
Should these be sorted?  I see inconsistencies in the order of the screenshot which is a bit confusing.